### ISSUE TITLE
Build as Java 9 modules

### ISSUE THREAD

**Issue by gnodet** [jline/jline3#70]

**Comment by gnodet** [comment]
I have an old branch I started earlier: https://github.com/jline/jline3/tree/wip-jigsaw
I'm not really keen on integrating those changes with all the repackaged dependencies, I'd rather wait until the dependencies have their own metadata.
That said, if you want to provide a PR with the automatic module names, I'd happily merge it.

**Comment by Thihup** [comment]
Would now be possible to add the module-info to the project?
There are two ways: requiring a JDK 9+ to build the project and adding the module-info.java or using a tool like Moditect to generate the module-info and use JDK 8 to build.

**Comment by ice1000** [comment]
It would be very nice if this can natively support jpms, so downstream projects can use jlink.

**Comment by bowbahdoe** [comment]
Bumping. Downstream dependencies like https://github.com/ctongfei/progressbar still need a module-info from this in order to be usable with jlink

**Comment by gnodet** [comment]
> Bumping. Downstream dependencies like https://github.com/ctongfei/progressbar still need a module-info from this in order to be usable with jlink

Do you think you could rais a PR for that ?

**Comment by bowbahdoe** [comment]
@gnodet I looked into adding module descriptors (first by locally bumping to java 9 and adding them and translating to moditect after), but I ran into a few issues

1. There are optional dependencies like jansi which are not themselves modularized (the whole tree needs to be modularized for people to not get hit with an error). 

I started on that work a bit, https://github.com/fusesource/jansi/pull/254.

2. What to do about the "uberjar" that is org.jline/jline with regard to the submodules which are available separately?

https://mvnrepository.com/artifact/org.jline/jline

**Comment by bowbahdoe** [comment]
And I suppose also

3. I'm not sure if I'm aware of all the service provider usages. (those go in the module-info too)

**Comment by gnodet** [comment]
> @gnodet I looked into adding module descriptors (first by locally bumping to java 9 and adding them and translating to moditect after), but I ran into a few issues
> 
> 1. There are optional dependencies like jansi which are not themselves modularized (the whole tree needs to be modularized for people to not get hit with an error).
> 
> I started on that work a bit, [fusesource/jansi#254](https://github.com/fusesource/jansi/pull/254).

Yes, that's the problem, all dependencies have to be modularized... I'll have a look at your PR.

> 2. What to do about the "uberjar" that is org.jline/jline with regard to the submodules which are available separately?

I think we need to ignore that one.  If people want to use java modules, they'll have to use the thinker jars.

For services, the TerminalProvider is obviously the main one.  Implementations are provided by jline-terminal, jline-terminal-jna and jline-terminal-jansi.

**Comment by garretwilson** [comment]
@gnodet thank you for working on this.

> See the [JPMS documentation](https://jline.org/docs/modules/jpms) for details.

It looks like that link is broken.
